trivial-builders.nix: Add input argument `passthru` to makeSetupHook

ShamrockLee opened the pull request:
This patch add an input argumnet `passthru` to `makeSetupHook`, and enable users to specify `passthru` attributes the way they specify the `meta` attributes in `makeSetupHook`

One significant use case is adding `passthru.tests` to setup-hooks,
and help increase test coverage for mission-critical setup-hooks.

The Nix expression of `postgresqlTestHook` has been updated to make use of the passthru attirbute. (No overrideAttrs are needed now.) This should evaluates to the same output as before.

`wrapGappsHook` could also be updated. However, as it currently passes the `passthru` through `substitutions` (weird enough), the modification would change the derivation hash, resulting in a large-scale rebuild. A successive PR will be made targeting the staging branch as this one gets merged.

Cc: @roberth `postgresqlTestHook` maintainer

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`) (Nothing has changed.)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ShamrockLee committed:
trivial-builders.nix: Add input argument `passthru` to makeSetupHook

One significant use case is adding `passthru.tests` to setup-hooks,
and help increase test coverage for mission-critical setup-hooks.

As `meta`, `passthru` doesn't go into the build script directly.
However, passing an empty set to `passthru` breaks nixpkgs-review
and OfBorg tests, so pass it only when specified.

ShamrockLee commented:
Just notice that passing `passthru = { };` breaks nixpkgs-review's and OfBorg's evaluation of outputs.

Now it is only passed when specified by the user.

ShamrockLee commented:
@ofborg build postgresqlTestHook postgresqlTestHook.passthru.tests

roberth committed:
setup hooks: substitutions.passthru -> passthru

roberth committed:
makeSetupHook: Deprecate substitutions.passthru

roberth commented:
I think I've found the problem.

@ofborg build postgresqlTestHook postgresqlTestHook.passthru.tests firefox-esr gnome3.zenity

ShamrockLee commented:
I see it now. Some hooks ends up specifying  `passthru` inside `substitute`, causing conflict when another `passthru = { }` is specified by the inheritance of the default value.

ShamrockLee commented:
I see it now. Some hooks ends up specifying  `passthru` inside `substitutions`, causing conflict when another `passthru = { }` is specified by the inheritance of the default value.

To deprecate the `substitutions.passthru`, we'll have to correct all the hooks using such pattern. If the `substitutions` attribute set contains nothing else, the build script of the hook will change, leading to the change of the derivation hash. One such hook is `wrapGApps`, the change of which will definitely trigger a mass rebuild.

ShamrockLee commented on pkgs/build-support/trivial-builders.nix:
```diff
@@ -523,12 +523,26 @@ rec {
         inherit meta;
         strictDeps = true;
+        # TODO 2023-01, no backport: inherit passthru
+        passthru = args.passthru or (
```
This way, the `{ passthru ? { } }:` will have nothing to do with the actual `passthru` passed into the derivation.

How about this
```
passthru = passthru // lib.optionalAttrs (substitutions?passthru)
  (lib.warn "makeSetupHook (name = ${lib.strings.escapeNixString name}): `substitutions.passthru` is deprecated. Please set `passthru` directly." substitutions.passthru);
```

roberth committed:
wrapGAppsHook: Set name

The previous commit, removing substitutions, causes a mass rebuild,
so we use the opportunity to set the name; also a mass rebuild.

roberth commented:
Hmm, I didn't expect a mass rebuild, but like you say, I've changed the outcome of `lib.optionalString (substitutions != {})`.


> To deprecate the `substitutions.passthru`, we'll have to correct all the hooks using such pattern.

That's what I did.
`wrapGAppsHook` is the only one for which builder changes. I've checked that the hook script does not rely on unspecified variables to be substituted, so this change should not pose a problem.

I'll change the branch to staging.

ShamrockLee commented:
The Ofborg failure is probably a `wrapGAppsHook.passthru.tests` issue, as this PR should not even change the build process.

ShamrockLee commented:
@roberth @Ericson2314 Anything else I could help to get it merge?
